File: windows-app/Plantoir.Core/Assist/AssistModelStore.cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Plantoir.Core.Scripting;

namespace Plantoir.Core.Assist;

/// <summary>What the store is doing, in words a teacher can read. Mirrors the mac's `AssistModelStore.State`.</summary>
public enum AssistModelStoreState
{
    Missing,
    Downloading,
    Ready,
    Failed,
}

/// <summary>
/// Where the local model lives on this PC, and how it gets here.
///
/// The weights are NOT bundled with the app — a 1–2.5 GB model in every
/// install would have to be downloaded by every teacher on every update,
/// landing on PCs whose teacher never opens the assistant. It is fetched
/// once, into <c>%LOCALAPPDATA%\Plantoir\models</c>, and stays there across
/// app updates. Mirrors the mac's `AssistModelStore.swift`.
/// </summary>
public sealed class AssistModelStore
{
    public AssistModelTier Tier { get; }

    public AssistModelStoreState State { get; private set; }

    /// <summary>How far a download has got — 0 outside <see cref="AssistModelStoreState.Downloading"/>.</summary>
    public double FractionComplete { get; private set; }
    public long ReceivedBytes { get; private set; }
    public long TotalBytes { get; private set; }

    /// <summary>Set only in <see cref="AssistModelStoreState.Failed"/>.</summary>
    public string? FailureReason { get; private set; }

    /// <summary>Raised after every state change, so a view can redraw without polling.</summary>
    public event Action? Changed;

    private readonly LocalModel _model;
    private CancellationTokenSource? _cancellation;

    public AssistModelStore(AssistModelTier tier)
    {
        Tier = tier;
        _model = new LocalModel(tier);
        State = _model.IsInstalled() ? AssistModelStoreState.Ready : AssistModelStoreState.Missing;
    }

    /// <summary>Is the model already here, and the right size?</summary>
    public bool IsReady => _model.IsInstalled();

    /// <summary>
    /// Fetch the weights, reporting progress. Does nothing when they are
    /// already here, or a download is already running — the second call is
    /// what a re-opened Settings window or a second assistant window
    /// produces, and it must find the SAME download in flight, not start a
    /// competing one onto the same file.
    /// </summary>
    public void Download()
    {
        if (IsReady)
        {
            State = AssistModelStoreState.Ready;
            Notify();
            return;
        }
        if (State == AssistModelStoreState.Downloading) return;

        _cancellation = new CancellationTokenSource();
        State = AssistModelStoreState.Downloading;
        FractionComplete = 0;
        ReceivedBytes = 0;
        TotalBytes = Tier.DownloadBytes();
        FailureReason = null;
        Notify();

        ActivityTrail.Note(ActivityTrail.Event.AssistantModelDownloadStarted,
            $"started downloading {Tier.DisplayName()} — {Tier.DownloadDescription()}");

        var progress = new Progress<LocalModel.Fetching>(state =>
        {
            ReceivedBytes = state.Bytes;
            TotalBytes = state.Known ? state.Total : Tier.DownloadBytes();
            FractionComplete = state.Known ? state.Percent / 100.0 : 0;
            Notify();
        });

        var token = _cancellation.Token;
        _ = RunDownload(progress, token);
    }

    private async Task RunDownload(IProgress<LocalModel.Fetching> progress, CancellationToken token)
    {
        // LocalModel.Install catches everything internally, cancellation
        // included, and answers false rather than throwing — so there is
        // nothing here to catch. token.IsCancellationRequested is the only
        // way to tell "the teacher stopped it" (Cancel() already handled
        // the state) apart from "it genuinely failed" (report why).
        bool installed = await _model.Install(progress, token).ConfigureAwait(false);
        if (token.IsCancellationRequested) return;

        if (installed)
        {
            State = AssistModelStoreState.Ready;
            ActivityTrail.Note(ActivityTrail.Event.AssistantModelDownloaded,
                $"finished downloading {Tier.DisplayName()} — {Tier.DownloadDescription()}");
        }
        else
        {
            FailureReason = "The download didn’t finish. Check the network and try again.";
            State = AssistModelStoreState.Failed;
            ActivityTrail.Note(ActivityTrail.Event.AssistantModelDownloadFailed,
                $"could not download {Tier.DisplayName()} — {FailureReason}");
        }
        Notify();
    }

    /// <summary>Stop a download in flight — the teacher closed the window, or pressed Stop.</summary>
    public void Cancel()
    {
        if (State != AssistModelStoreState.Downloading) return;
        _cancellation?.Cancel();
        _cancellation = null;
        State = AssistModelStoreState.Missing;
        ActivityTrail.Note(ActivityTrail.Event.AssistantModelDownloadStopped,
            $"stopped downloading {Tier.DisplayName()}");
        Notify();
    }

    /// <summary>
    /// Why this model cannot be removed right now, in a teacher's words, or
    /// null when it can be. Naming the section rather than saying "not
    /// available" is the same rule <c>AssistActivity</c> follows elsewhere:
    /// somebody told they cannot do a thing needs to know what to close.
    ///
    /// The guard is deliberately about ANY open assistant, not about whether
    /// that window happens to be using this particular rung — mirrors the
    /// mac's `AssistModelLibrary.reasonItCannotBeRemoved`. The window loaded
    /// whatever was chosen when it opened, and the choice can have changed
    /// since; working out which file is genuinely mapped into a running
    /// `llama-server.exe` means tracking state this app does not keep, and
    /// getting it wrong deletes the weights out from under a running engine.
    /// </summary>
    public string? ReasonItCannotBeRemoved()
    {
        if (!IsReady) return null;
        if (AssistActivity.Active is not { } active) return null;
        return $"Close the assistant for {active.CourseCode} Section {active.SectionNumber} " +
               "before removing this.";
    }

    /// <summary>Whether the Remove button should do anything right now.</summary>
    public bool MayRemove()
    {
        if (!IsReady) return false;
        return ReasonItCannotBeRemoved() is null;
    }

    /// <summary>
    /// Delete the weights to get the space back. Does nothing when it is not
    /// allowed, so a stale view cannot remove what the panel says it cannot —
    /// see <see cref="MayRemove"/>.
    /// </summary>
    public void Remove()
    {
        if (!MayRemove()) return;
        Cancel();
        try
        {
            string path = LocalModel.GetModelPath(Tier);
            if (File.Exists(path)) File.Delete(path);
            State = AssistModelStoreState.Missing;
            ActivityTrail.Note(ActivityTrail.Event.AssistantModelRemoved,
                $"removed {Tier.DisplayName()} — {Tier.DownloadDescription()} freed");
        }
        catch (Exception error)
        {
            FailureReason = $"Could not remove it: {error.Message}";
            State = AssistModelStoreState.Failed;
        }
        Notify();
    }

    /// <summary>What this tier takes on disk right now, or null when it is not here.</summary>
    public static long? BytesOnDisk(AssistModelTier tier)
    {
        string path = LocalModel.GetModelPath(tier);
        if (!File.Exists(path)) return null;
        return new FileInfo(path).Length;
    }

    private void Notify() => Changed?.Invoke();
}

/// <summary>
/// One store per rung, for the whole app.
///
/// The bug this exists to stop is a double download: a teacher who presses
/// Download in Settings and then opens the assistant (or vice versa) while
/// it is running must find the SAME download in progress, not a second one
/// racing it onto the same file — the failure mode this project has already
/// paid for once on the mac side. Mirrors `AssistModelStores.swift`.
/// </summary>
public static class AssistModelStores
{
    private static readonly System.Collections.Generic.Dictionary<AssistModelTier, AssistModelStore> Stores = new();
    private static readonly object Gate = new();

    /// <summary>The one store for this rung, made on first use.</summary>
    public static AssistModelStore Store(AssistModelTier tier)
    {
        lock (Gate)
        {
            if (Stores.TryGetValue(tier, out var existing)) return existing;
            var made = new AssistModelStore(tier);
            Stores[tier] = made;
            return made;
        }
    }

    /// <summary>Forget everything, for tests — process-wide state that outlives a test answers the next test's questions.</summary>
    public static void Reset()
    {
        lock (Gate)
        {
            foreach (var store in Stores.Values) store.Cancel();
            Stores.Clear();
        }
    }
}

File: mac-app/QuartzTeachers/Models/Assist/AssistModelStore.swift
import Foundation
import Observation

/// Where the local model lives on this Mac, and how it gets here.
///
/// The weights are NOT in the app bundle. A 1–5 GB model inside the app would
/// have to be downloaded by every teacher on every update, signed and
/// notarised each time, and would land on Macs whose teacher never opens the
/// assistant. It is fetched once, into Application Support, and stays there
/// across app updates.
@Observable
final class AssistModelStore {

    // MARK: - Types

    /// What the store is doing, in words a teacher can read.
    enum State: Equatable {
        case missing
        case downloading(fractionComplete: Double, receivedBytes: Int64, totalBytes: Int64)
        case ready
        case failed(reason: String)
    }

    // MARK: - Stored properties

    /// The model this Mac should run.
    let tier: AssistModelTier

    /// Where the store is up to.
    private(set) var state: State = .missing

    /// The download in flight, so it can be cancelled with the window.
    private var task: URLSessionDownloadTask?

    /// Kept alive for the length of a download; its delegate reports progress.
    private var session: URLSession?

    // MARK: - Computed properties

    /// Somewhere else to keep the weights, for tests only.
    ///
    /// The same seam `ActivityTrail.store` has, and for a sharper reason:
    /// these tests DELETE files. Without a way to point the store at a
    /// temporary folder, a test of "removing frees the space" would remove
    /// the weights off the machine it was running on — several gigabytes the
    /// developer then re-downloads, having been given no reason to connect it
    /// to a green test run.
    nonisolated(unsafe) static var directoryOverride: URL?

    /// `~/Library/Application Support/Plantoir/models`.
    ///
    /// **Under the unit suite, the suite's throwaway home instead** (issue
    /// #264), through `RealHome.forFiles`. Three tests used to stat the real
    /// weights here, so a panel sentence they checked depended on what the
    /// Mac running them had downloaded — a suite that answers differently on
    /// a Mac with weights. The app a UI test drives has no XCTest in it and
    /// keeps the real folder, which `AssistantRolloverUITests` needs: it runs
    /// against a real model.
    static var directoryURL: URL {
        if let override = directoryOverride {
            return override
        }
        return RealHome.forFiles
            .appendingPathComponent("Library", isDirectory: true)
            .appendingPathComponent("Application Support", isDirectory: true)
            .appendingPathComponent("Plantoir", isDirectory: true)
            .appendingPathComponent("models", isDirectory: true)
    }

    /// Where this tier's weights sit once downloaded.
    var fileURL: URL {
        return AssistModelStore.directoryURL.appendingPathComponent(tier.fileName)
    }

    /// Is the model already here, and the right size? A truncated download —
    /// a closed lid, a dropped network — leaves a file that exists and is
    /// useless, and llama.cpp's failure on one is not obviously about size.
    var isReady: Bool {
        let path: String = fileURL.path
        guard FileManager.default.fileExists(atPath: path) else {
            return false
        }
        let attributes: [FileAttributeKey: Any]? = try? FileManager.default.attributesOfItem(atPath: path)
        let size: Int64 = (attributes?[.size] as? NSNumber)?.int64Value ?? 0
        return size == tier.downloadBytes
    }

    // MARK: - Initializer

    init(tier: AssistModelTier) {
        self.tier = tier
        if isReady {
            state = .ready
        }
    }

    // MARK: - Functions

    /// Fetch the weights, reporting progress. Does nothing when they are
    /// already here.
    func download() {
        if isReady {
            state = .ready
            return
        }
        if case .downloading = state {
            return
        }

        do {
            try FileManager.default.createDirectory(
                at: AssistModelStore.directoryURL, withIntermediateDirectories: true
            )
        } catch {
            state = .failed(reason: "Could not make a place for the model: \(error.localizedDescription)")
            return
        }

        // A part-finished file from a previous attempt is never resumed: the
        // ranges would have to match exactly and a wrong guess produces a
        // corrupt model that fails much later, somewhere less obvious.
        try? FileManager.default.removeItem(at: fileURL)

        state = .downloading(fractionComplete: 0, receivedBytes: 0, totalBytes: tier.downloadBytes)
        ActivityTrail.note(
            .assistantModelDownloadStarted,
            "started downloading \(tier.displayName) — \(tier.downloadDescription)"
        )

        let delegate: DownloadDelegate = DownloadDelegate(store: self)
        let configuration: URLSessionConfiguration = .default
        configuration.waitsForConnectivity = true
        let session: URLSession = URLSession(configuration: configuration, delegate: delegate, delegateQueue: nil)
        self.session = session

        let task: URLSessionDownloadTask = session.downloadTask(with: tier.downloadURL)
        self.task = task
        task.resume()
    }

    /// Delete the weights to get the space back.
    ///
    /// Deliberately NOT hidden behind "are you sure it is not in use" logic
    /// here: whether it is safe to remove is a question about what windows
    /// are open, which this type cannot see and `AssistModelLibrary` can.
    /// This does the deletion and nothing else, so a test can drive it
    /// without standing an assistant window up.
    ///
    /// Removing the model a teacher is currently set to use is allowed on
    /// purpose — it is not a broken state, it is the state every Mac is in
    /// before the first download, and opening the assistant offers to fetch
    /// it again.
    func remove() {
        cancel()
        do {
            try FileManager.default.removeItem(at: fileURL)
        } catch {
            let missing: Bool = (error as NSError).code == NSFileNoSuchFileError
            if !missing {
                state = .failed(reason: "Could not remove it: \(error.localizedDescription)")
                return
            }
        }
        state = .missing
        ActivityTrail.note(
            .assistantModelRemoved,
            "removed \(tier.displayName) — \(tier.downloadDescription) freed"
        )
    }

    /// What this tier takes on disk right now, or nil when it is not here.
    ///
    /// Reads the actual file rather than reporting `downloadBytes`, so a
    /// half-finished download is described by its real size instead of by the
    /// size it was going to be.
    static func bytesOnDisk(for tier: AssistModelTier) -> Int64? {
        let url: URL = AssistModelStore.directoryURL.appendingPathComponent(tier.fileName)
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: url.path) else {
            return nil
        }
        return (attributes[.size] as? NSNumber)?.int64Value
    }

    /// Stop a download in flight — the teacher closed the window.
    func cancel() {
        task?.cancel()
        task = nil
        session?.invalidateAndCancel()
        session = nil
        if case .downloading = state {
            state = .missing
            ActivityTrail.note(
                .assistantModelDownloadStopped,
                "stopped downloading \(tier.displayName)"
            )
        }
    }

    /// Called by the delegate as bytes arrive.
    fileprivate func report(received: Int64, expected: Int64) {
        let total: Int64 = expected > 0 ? expected : tier.downloadBytes
        var fraction: Double = 0
        if total > 0 {
            fraction = Double(received) / Double(total)
        }
        state = .downloading(fractionComplete: fraction, receivedBytes: received, totalBytes: total)
    }

    /// Called by the delegate when the body has landed in a temporary file.
    fileprivate func finish(movingFrom temporary: URL) {
        do {
            try? FileManager.default.removeItem(at: fileURL)
            try FileManager.default.moveItem(at: temporary, to: fileURL)
        } catch {
            state = .failed(reason: "Could not save the model: \(error.localizedDescription)")
            return
        }

        // Trust the bytes on disk rather than the fact that the transfer
        // ended: a proxy or a captive portal answers 200 with something that
        // is not a model at all.
        if isReady {
            state = .ready
            ActivityTrail.note(
                .assistantModelDownloaded,
                "finished downloading \(tier.displayName) — \(tier.downloadDescription)"
            )
        } else {
            try? FileManager.default.removeItem(at: fileURL)
            let reason: String = "The download finished but the file is the wrong size. Try again."
            state = .failed(reason: reason)
            ActivityTrail.note(
                .assistantModelDownloadFailed,
                "could not download \(tier.displayName) — \(reason)"
            )
        }
        session?.finishTasksAndInvalidate()
        session = nil
        task = nil
    }

    fileprivate func fail(_ message: String) {
        state = .failed(reason: message)
        ActivityTrail.note(
            .assistantModelDownloadFailed,
            "could not download \(tier.displayName) — \(message)"
        )
        session?.finishTasksAndInvalidate()
        session = nil
        task = nil
    }
}

// MARK: - Download delegate

/// URLSession calls back on its own queue, so every hop into the store is
/// bounced to the main actor where the observable state lives.
private final class DownloadDelegate: NSObject, URLSessionDownloadDelegate, @unchecked Sendable {

    // MARK: - Stored properties

    private let store: AssistModelStore

    // MARK: - Initializer

    init(store: AssistModelStore) {
        self.store = store
    }

    // MARK: - Functions

    func urlSession(_ session: URLSession,
                    downloadTask: URLSessionDownloadTask,
                    didWriteData bytesWritten: Int64,
                    totalBytesWritten: Int64,
                    totalBytesExpectedToWrite: Int64) {
        let received: Int64 = totalBytesWritten
        let expected: Int64 = totalBytesExpectedToWrite
        Task { @MainActor in
            store.report(received: received, expected: expected)
        }
    }

    func urlSession(_ session: URLSession,
                    downloadTask: URLSessionDownloadTask,
                    didFinishDownloadingTo location: URL) {
        // The temporary file is deleted the moment this method returns, so it
        // is moved somewhere of our own SYNCHRONOUSLY here, and only then
        // handed to the main actor.
        let holding: URL = FileManager.default.temporaryDirectory
            .appendingPathComponent("plantoir-model-\(UUID().uuidString).gguf")
        do {
            try FileManager.default.moveItem(at: location, to: holding)
        } catch {
            let message: String = error.localizedDescription
            Task { @MainActor in
                store.fail("Could not save the model: \(message)")
            }
            return
        }
        Task { @MainActor in
            store.finish(movingFrom: holding)
        }
    }

    func urlSession(_ session: URLSession,
                    task: URLSessionTask,
                    didCompleteWithError error: Error?) {
        guard let error else {
            return
        }
        let cancelled: Bool = (error as NSError).code == NSURLErrorCancelled
        if cancelled {
            return
        }
        let message: String = error.localizedDescription
        Task { @MainActor in
            store.fail(message)
        }
    }
}
